### PR TITLE
paginate by startkey

### LINKED ISSUE #436: proposal: a faster result iteration
Please [read these guidelines](http://ibm.biz/cdt-issue-guide) before opening an issue.

## Bug Description

Iterating over a large dataset is sometimes slow. This is not a library disfunction but is related to how couchdb works, ie.

```sh
# take an acceptable time
SKIP=0
time curl -sH 'Content-Type: application/json' \
    http://localhost:5984/db/_design/dd/_view/vv\?limit\=500\&skip\=$SKIP\&group\=true

# this will take ages
SKIP=820000
time curl -sH 'Content-Type: application/json' \
    http://localhost:5984/db/_design/dd/_view/vv\?limit\=500\&skip\=$SKIP\&group\=true

```

So this is inherently true for your library when using current `Result` class logic. 
I've noticed that the different pagination approach [manual propose](https://docs.couchdb.org/en/stable/ddocs/views/pagination.html#paging-alternate-method) does not incur in this slowdown.

I drafted a switch implementation that can allow to iterate based on `startkey` instead of `skip` and you can [see it here](https://github.com/cloudant/python-cloudant/compare/master...aogier:feature/faster-iterator?expand=1), as you see is quite simple and yes is badly coded :grimacing: because is a PoC and I left original implementation under the new alternative.

What do you think? This is maybe somehow related to #389 or not, anyway for me it's a complete game changer that I'd like to have upstream.

Thank you, regards

### PR DESCRIPTION
**Pull request by aogier**
## Checklist

- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](../blob/master/DCO1.1.txt)
- [x] Added tests for code changes _or_ test/build only changes
- [x] Updated the change log file (`CHANGES.md`|`CHANGELOG.md`) _or_ test/build only changes
- [x] Completed the PR template below:

## Description

Make `Result`'s iteration faster paginating by `startkey` instead of `skip`.

Fixes #436 

## Approach

As per #436 , paginating by `skip` could be super-slow. This patch fix the problem by using `startkey` so every iteration run in constant time.

## Schema & API Changes

- "No change"

## Security and Privacy

- "No change"

## Testing

- No new tests because current ones are enough

## Monitoring and Logging

- "No change"

### CODE REVIEW

File: src/cloudant/result.py [hunk 1]
```diff
@@ -347,24 +347,27 @@ def __iter__(self):
...
+                startkey=last,
                 **self.options
             )
             result = self._parse_data(response)
-            skip += self._page_size
             if result:
+                last = result.pop()
                 for row in result:
                     yield row
                 if len(result) < self._page_size:
```
Comment (smithsz) [inline src/cloudant/result.py]: Change to `if len(result) < self._page_size + 1`?

Also, I think a comment here would be good. Something like:

> We expect `len(result) = self._page_size + 1` results, if we have ` self._page_size` or less it means we are on the last page and need to return the last result.

**Comment by aogier** [comment]
I've found a bug while implementing comments requests: when a view's key is an integer, view does not return because startkey='' apparently want to cast to string:

```sh
# this does work
c /temp-cloudant/_design/dd/_view/vv\?limit=3                
{
  "total_rows": 6,
  "offset": 0,
  "rows": [
    {
      "id": "f41a9664608b060ce1a488ed96005a35",
...

# this does not

c /temp-cloudant/_design/dd/_view/vv\?limit=3\&startkey\='""'
{
  "total_rows": 6,
  "offset": 6,
  "rows": []
}

```

File: src/cloudant/result.py (hunk 1 above)
Comment (aogier) [reply src/cloudant/result.py]: you're right, this way we save a hit on the worst case when ndocs%pagesize==0 ... neat !

**Comment by aogier** [comment]
Ok I've changed a bit the approach so we use `startkey` in GET if and only if user specifies one. I've implemented this via a partial object (I like them so much, so elegant). Also, added the comment and avoid an empty hit if we are at the view's end!

**Comment by aogier** [comment]
I'm working on tests to cover the integer key corner case

**Comment by aogier** [comment]
added a test that would not have passed in 19a4abe078f9ba55083ed44fc5baddf18b45437c but pass now in 15d216d46fb68a0f6e7b43089a73a5a6d4324996 ... reverse TDD :+1:

**Comment by aogier** [comment]
have to write more tests

**Comment by aogier** [comment]
ok I've added another test for observing pagination behaviour and now everything seems pretty stable and robust to me. @smithsz thank you for letting me know what you think :)

ciao

File: src/cloudant/result.py [hunk 2]
```diff
@@ -327,7 +328,7 @@ def _handle_result_by_key_slice(self, key_slice):
     def __iter__(self):
         """
         Provides iteration support, primarily for large data collections.
-        The iterator uses the ``skip`` and ``limit`` options to consume
+        The iterator uses the ``startkey`` and ``limit`` options to consume
```
Comment (smithsz) [inline src/cloudant/result.py]: and ``startkey_docid``

File: tests/unit/result_tests.py [hunk 3]
```diff
@@ -643,5 +645,43 @@ def test_iteration_no_data(self):
         result = Result(self.view001, startkey='ruby')
         self.assertEqual([x for x in result], [])
 
+    def test_iteration_integer_keys(self):
+        """
+        Test that iteration works as expected when keys are integer.
+        """
+        result = Result(self.view007)
```
Comment (smithsz) [inline tests/unit/result_tests.py]: can you put a small page size in here? maybe `page_size=10`

File: src/cloudant/result.py (hunk 2 above)
Comment (aogier) [reply src/cloudant/result.py]: ah ok, I've left it back on purpose viewing it as an "hidden feature" but I'll amend docs

File: tests/unit/result_tests.py (hunk 3 above)
Comment (aogier) [reply tests/unit/result_tests.py]: sure

File: tests/unit/result_tests.py [hunk 4]
```diff
@@ -643,5 +645,43 @@ def test_iteration_no_data(self):
...
+        """
+
+        class CallMock:
+            last_call = None
+
+            def __init__(self, outer):
+                self.outer = outer
+
+            def call(self, *args, **kwargs):
+                self.outer.assertNotEqual((args, kwargs), self.last_call,
```
Comment (smithsz) [inline tests/unit/result_tests.py]: This is just asserting that the current view query isn't the same as the previous right? How about checking against a list of expected stuff, i.e.
```
call 1: {'limit': 28}
call 2: {'limit': 28, 'startkey': 1, 'startkey_docid': 'julia027'}
call 3: {'limit': 28, 'startkey': 1, 'startkey_docid': 'julia054'}
call 4: {'limit': 28, 'startkey': 1, 'startkey_docid': 'julia081'}
```

**Comment by smithsz** [comment]
This is a great. I've been doing some testing with other options (i.e. descending=True) and everything works nicely.
I notice there's a check in there for `limit` and `skip` params not being supported. I guess this is still true for `limit` but we should be able to use `skip` now right? It'll want removing from the query params after the first call though. What do you think?

File: tests/unit/result_tests.py (hunk 4 above)
Comment (aogier) [reply tests/unit/result_tests.py]: It's indeed more strict, I like it

**Comment by aogier** [comment]
Seems a good idea to me, it opens up to more use cases and it's totally manageable in the current implementation. 

BTW, I think we could also reason about allowing `limit`: we use `limit` in the request context for efficiently paginating over a result set, but from the user standpoint a `limit` clause mean "I want to iterate over this set and get at most N total results" and this could totally be feasible by servin count(results) < limit. What do you think? Just a couple cents about another iteration eh :)

I'll begin to implement for a skip param!

File: src/cloudant/result.py [hunk 5]
```diff
@@ -347,21 +348,47 @@ def __iter__(self):
         except ValueError:
             raise ResultException(104, self._page_size)
 
-        skip = 0
+        initial_options = {option: value
```
Comment (smithsz) [inline src/cloudant/result.py]: Not sure about this. Bit tricky to read. How about...
```py
        _call = partial(self._ref, limit=self._page_size + 1, **self.options)

        init_opts = {
            'skip': self.options.pop('skip', None),
            'startkey': self.options.pop('startkey', None)
        }

        response = _call(**{k: v for k, v in init_opts.items() if v is not None})
```

Comment (aogier) [reply src/cloudant/result.py]: ok, it's more concise and definitely easier to read!

**Comment by aogier** [comment]
did the last changes and also added a changelog entry!

**Comment by aogier** [comment]
queries are to be paginated via `bookmark`, I'm adding more tests

**Comment by aogier** [comment]
ok so basically I was abusing nose's filtering options and skipped query tests, now there is a possibile implementation on which I decided to stay straight and implement a specific inner method for queries and views as they are so different in schema. Thank you, regards

File: tests/unit/query_result_tests.py [hunk 6]
```diff
@@ -428,21 +428,11 @@ def test_iteration_with_invalid_options(self):
         Test that iteration raises an exception when "skip" and/or "limit" are
```
Comment (emlaver) [inline tests/unit/query_result_tests.py]: Need to fix this comment now that tests with skip are removed.

**Comment by emlaver** [comment]
For tests to run under Python 3, this will need to be rebased to include #438 which fixes the `pylint` warning.

File: tests/unit/query_result_tests.py (hunk 6 above)
Comment (aogier) [reply tests/unit/query_result_tests.py]: you're right!

**Comment by aogier** [comment]
ok I've fixed test docstring and rebased on 2eb1261838d02b5ee63d4a572095a78bcfcee825, were jenkins failing because of pylint?

**Comment by emlaver** [comment]
@aogier Yes, the Python 3 tests were showing the pylint warning.

File: CHANGES.md [hunk 7]
```diff
@@ -1,5 +1,6 @@
 # Unreleased
 
+- [IMPROVED] Faster `Result` iterator.
```
Comment (emlaver) [inline CHANGES.md]: We should probably give some detail and not use 'Faster'
i.e. `Improve Result iteration by paginating with 'startkey'`.  @smithsz wdyt?

**Comment by aogier** [comment]
Great !

I've squashed it and modified the changelog entry:
```
- [IMPROVED] Updated `Result` iteration by paginating with views' `startkey` and
  query's `bookmark`.
```

Thank you, regards
